### PR TITLE
add mergeLimit function as mentioned in #374

### PR DESCRIPTION
**Pull request by dayoadeyemi**

### CODE REVIEW

**Comment by vqvu** [comment]
This breaks backpressure. You should be using modified version of the current implementation of `merge`. Then have `merge` return `this.mergeLimit(Infinity)`.

You all so need to add a `noValueOnErrorTest`. See the other tests for details.

File: lib/index.js [hunk 1]
```diff
@@ -3611,6 +3611,78 @@ Stream.prototype.merge = function () {
...
+ * @name Stream.mergeLimit()
+ * @api public
+ *
+ * var txt = _(['foo.txt', 'bar.txt']).map(readFile)
+ * var md = _(['baz.md']).map(readFile)
+ * var js = _(['bosh.js']).map(readFile)
+ *
+ * _([txt, md, js]).mergeLimit(2);
+ * // => contents of foo.txt, bar.txt, baz.txt and bosh.js in the order
+ * they were read, but bosh is not read until the others have comleted
```
Comment (vqvu) [inline lib/index.js]: Typo in "comleted". Also, this wording suggests bosh won't be read until both bar and baz have completed. I suggest "but bosh is not read until either foo/bar or baz has finished.

File: lib/index.js [hunk 2]
```diff
@@ -3611,6 +3611,78 @@ Stream.prototype.merge = function () {
...
+ * values for each stream end up in the merged stream. Values in the
+ * merged stream will, however, respect the order they were emitted from
+ * their respective streams.
+ *
+ * @id mergeLimit
+ * @section Higher-order Streams
+ * @name Stream.mergeLimit()
+ * @api public
+ *
+ * var txt = _(['foo.txt', 'bar.txt']).map(readFile)
```
Comment (vqvu) [inline lib/index.js]: These all need to be `flatMap(readFile)`.

File: lib/index.js [hunk 3]
```diff
@@ -3611,6 +3611,78 @@ Stream.prototype.merge = function () {
...
+ * running at any one time.
+ *
+ * Note that no guarantee is made with respect to the order in which
+ * values for each stream end up in the merged stream. Values in the
+ * merged stream will, however, respect the order they were emitted from
+ * their respective streams.
+ *
+ * @id mergeLimit
+ * @section Higher-order Streams
+ * @name Stream.mergeLimit()
```
Comment (vqvu) [inline lib/index.js]: Call this `Stream.mergeLimit(n)`. You're also missing a description of the parameter `n`.

**Comment by vqvu** [comment]
Finally, let's call this `mergeWithLimit` to match the existing `batchWIthTimeOrCount`.

**Comment by dayoadeyemi** [comment]
Yes that did break back-pressure, Instead of modifying the merge function I've made it so that there is just a transform function before the merge which prevents the stream from pushing any values until something from the current working set has ended.

File: lib/index.js [hunk 4]
```diff
@@ -3611,6 +3611,78 @@ Stream.prototype.merge = function () {
...
+            }
+            else if (x === nil) {
+                push(null, nil);
+            }
+            else {
+                index++;
+                push(err, x);
+                // console.log('start', x.id)
+                    x.once('end', function(){
+                    // console.log('end', x.id)
```
Comment (vqvu) [inline lib/index.js]: Your indentation for `x.once` is off. Also, I think you want to do `index--` in the `end` handler, right?

**Comment by vqvu** [comment]
> I've made it so that there is just a transform function before the merge which prevents the stream from pushing any values until something from the current working set has ended.

Oh, that's a much better idea.

**Comment by dayoadeyemi** [comment]
Yea if I don't index-- when the substream ends it will end up only allowing 1 process to run at a time. I've added that in and renamed index to processCount so that it is more descriptive.

File: lib/index.js [hunk 5]
```diff
@@ -3611,6 +3611,79 @@ Stream.prototype.merge = function () {
...
+ * running at any one time.
+ *
+ * Note that no guarantee is made with respect to the order in which
+ * values for each stream end up in the merged stream. Values in the
+ * merged stream will, however, respect the order they were emitted from
+ * their respective streams.
+ *
+ * @id mergeLimit
+ * @section Higher-order Streams
+ * @name Stream.mergeWithLimit()
```
Comment (vqvu) [inline lib/index.js]: Call this `@name Stream.mergeWithLimit()`.

You also need to add a `@param n` field for the parameter.

File: lib/index.js [hunk 6]
```diff
@@ -3611,6 +3611,79 @@ Stream.prototype.merge = function () {
...
+ * @api public
+ *
+ * var txt = _(['foo.txt', 'bar.txt']).flatMap(readFile)
+ * var md = _(['baz.md']).flatMap(readFile)
+ * var js = _(['bosh.js']).flatMap(readFile)
+ *
+ * _([txt, md, js]).mergeWithLimit(2);
+ * // => contents of foo.txt, bar.txt, baz.txt and bosh.js in the order
+ * they were read, but bosh.js is not read until either foo.txt and bar.txt
```
Comment (vqvu) [inline lib/index.js]: This doesn't format correctly. Run `grunt docs` and open `docs/index.html` to see what I mean (don't commit it!).

Everything after `@api public` is treated as code, so you need a `//` before each of these lines.

File: lib/index.js [hunk 7]
```diff
@@ -3611,6 +3611,79 @@ Stream.prototype.merge = function () {
...
+
+Stream.prototype.mergeWithLimit = function (n){
+    var self = this;
+    var processCount = 0;
+    var waiting = false;
+    n = typeof n != 'number' ? Infinity : n;
+    if (n === Infinity) {
+        return this.merge();
+    } else if (n < 1) {
+        return _([]);
```
Comment (vqvu) [inline lib/index.js]: This should throw an error instead of returning an empty stream.

**Comment by vqvu** [comment]
I think this is ready as soon as you fix a few issues with the documentation.

**Comment by svozza** [comment]
A couple of stray `console.log` debug statements commented out but other than that this looks good to merge (pardon the pun).

**Comment by vqvu** [comment]
Cool. Commented `console.log` are fine, since they're all over the rest of the codebase anyway.

Thanks, @dayoadeyemi for the PR!

**Pull request merged**